x-pack/plugins/files/server/file_client/file_client.ts

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { FileKind } from '../../common/types';
+import type { FileMetadataClient } from './file_metadata_client';
+import type { BlobStorageClient } from '../blob_storage_service';
+import { enforceMaxByteSizeTransform } from './stream_transforms';
+
+/**
+ * Wraps the {@link FileMetadataClient} and {@link BlobStorageClient} client
+ * to provide basic file CRUD functionality.
+ */
+export class FileClient {
+  constructor(
+    private fileKindDescriptor: FileKind,
+    private readonly metadataClient: FileMetadataClient,
+    private readonly blobStorageClient: BlobStorageClient
+  ) {}
+
+  public get fileKind(): string {
+    return this.fileKindDescriptor.id;
+  }
+
+  public create: FileMetadataClient['create'] = async (arg) => {
+    return this.metadataClient.create(arg);
+  };
+
+  public get: FileMetadataClient['get'] = async (arg) => {
+    return this.metadataClient.get(arg);
+  };
+
+  public update: FileMetadataClient['update'] = (arg) => {
+    return this.metadataClient.update(arg);
+  };
+
+  public async delete({ id, hasContent }: { id: string; hasContent?: boolean }) {
+    if (hasContent) await this.blobStorageClient.delete(id);
+    return this.metadataClient.delete({ id });
+  }
```

Comment:
We should mark file as "DELETED" before actually deleting the blob.

```suggestion
  public async delete({ id, hasContent }: { id: string; hasContent?: boolean }) {
    await this.metadataClient.markAsDeleted(id);
    if (hasContent) await this.blobStorageClient.delete(id);
    return this.metadataClient.delete({ id });
  }
```

Reply:
At the moment the `FileClient` is not aware of state transitions like this. I was thinking consumers of `FileClient` would implement this themselves.

I'm happy to go with your suggestion though. I'd prefer to do that in a follow up PR though since it will involve some more changes. Let me know what you think!